project_task_materials_stock/models/project.py

```diff
@@ -0,0 +1,151 @@
+        self.stock_move_ids = self.env['stock.move'].browse(move_ids)
+
+    def _get_analytic_line(self):
+        line_ids = [line.analytic_line_id.id for line in self.material_ids
+                    if line.analytic_line_id]
+        self.analytic_line_ids = \
+            self.env['account.analytic.line'].browse(line_ids)
+
+    @api.depends('stock_move_ids.state')
+    def _check_stock_state(self):
+        if not self.stock_move_ids:
+            self.stock_state = 'pending'
+        elif self.stock_move_ids.filtered(lambda r: r.state == 'confirmed'):
+            self.stock_state = 'confirmed'
+        elif self.stock_move_ids.filtered(lambda r: r.state == 'assigned'):
+            self.stock_state = 'assigned'
+        elif self.stock_move_ids.filtered(lambda r: r.state == 'done'):
+            self.stock_state = 'done'
+
+    stock_move_ids = fields.One2many(
+        comodel_name='stock.move', compute='_get_stock_move',
+        string='Stock Moves')
+    analytic_line_ids = fields.One2many(
+        comodel_name='account.analytic.line', compute='_get_analytic_line',
+        string='Analytic Lines')
+    consume_material = fields.Boolean(related='stage_id.consume_material')
+    stock_state = fields.Selection(
+        [('pending', 'Pending'),
+         ('confirmed', 'Confirmed'),
+         ('assigned', 'Assigned'),
+         ('done', 'Done')], compute='_check_stock_state', string='Stock State')
+
+    @api.one
+    def unlink_stock_move(self):
+        for move in self.stock_move_ids:
+            if move.state == 'assigned':
+                move.do_unreserve()
+            if move.state in ['waiting', 'confirmed', 'assigned']:
+                move.state = 'draft'
+            move.unlink()
+
+    @api.one
+    def write(self, vals):
+        res = super(Task, self).write(vals)
+        if 'stage_id' in vals:
+            if self.consume_material:
+                self.material_ids.create_stock_move()
+                self.material_ids.create_analytic_line()
+            else:
+                self.unlink_stock_move()
+                self.analytic_line_ids.unlink()
+        return res
+
+    @api.multi
+    def action_assign(self):
+        self.stock_move_ids.action_assign()
+
+    @api.multi
+    def action_done(self):
+        self.stock_move_ids.action_done()
+        pass
+
+
+class ProjectTaskMaterials(models.Model):
+    _inherit = "project.task.materials"
+    stock_move_id = fields.Many2one(
+        comodel_name='stock.move', string='Stock Move')
+    analytic_line_id = fields.Many2one(
+        comodel_name='account.analytic.line', string='Analytic Line')
+
+    def _prepare_stock_move(self):
+        product = self.product_id
+        res = {
+            'product_id': product.id,
+            'name': product.partner_ref,
+            'state': 'confirmed',
+            'product_uom': product.uom_id.id,
+            'product_uos': product.uos_id and product.uos_id.id or False,
```

Comment:
If you are not putting the same UoM in both fields (uom and uos), you cannot put the same quantity without calculating the converted quantity. Please put here also `product.uom_id.id` for simplicity (you are not going to work with another UoM in this case I think). You can even mention this question in known issues: "You cannot use another UoM in tasks except the product stock one".